assembly/src/assembler/procedure_cache.rs

```diff
@@ -0,0 +1,62 @@
+use super::{btree_map::Entry, BTreeMap, Procedure, ProcedureId, RpoDigest};
+
+#[cfg(test)]
+use super::btree_map::Values;
+
+/// The [ProcedureCache] is responsible for caching [Procedure]s. It allows [Procedure]s to be
+/// fetched using both [ProcedureId] and [RpoDigest].
+#[derive(Debug, Default)]
+pub struct ProcedureCache {
+    proc_id_map: BTreeMap<ProcedureId, RpoDigest>,
+    mast_map: BTreeMap<RpoDigest, Procedure>,
+}
+
+impl ProcedureCache {
+    // ACCESSORS
+    // --------------------------------------------------------------------------------------------
+    /// Returns a [Procedure] reference corresponding to the [ProcedureId].
+    pub fn get_by_id(&self, id: &ProcedureId) -> Option<&Procedure> {
+        self.proc_id_map.get(id).and_then(|mast_hash| self.mast_map.get(mast_hash))
+    }
+
+    /// Returns a [Procedure] referene corresponding to the MAST root ([RpoDigest]).
+    pub fn get_by_hash(&self, root: &RpoDigest) -> Option<&Procedure> {
+        self.mast_map.get(root)
+    }
+
+    /// Returns true if the [ProcedureCache] contains a [Procedure] for the specified [ProcedureId].
+    pub fn contains_id(&self, id: &ProcedureId) -> bool {
+        self.proc_id_map.contains_key(id)
+    }
+
+    /// Returns true if the [ProcedureCache] contains a [Procedure] for the specified root
+    /// ([RpoDigest]).
+    pub fn contains_hash(&self, root: &RpoDigest) -> bool {
+        self.mast_map.contains_key(root)
+    }
+
+    /// Returns an iterator over the [Procedure]s in the [ProcedureCache].
+    #[cfg(test)]
+    pub fn values(&self) -> Values<RpoDigest, Procedure> {
+        self.mast_map.values()
+    }
+
+    /// Returns the number of [Procedure]s in the [ProcedureCache].
+    #[cfg(test)]
+    pub fn len(&self) -> usize {
+        self.mast_map.len()
+    }
+
+    // MUTATORS
+    // --------------------------------------------------------------------------------------------
+    /// Inserts a [Procedure] into the [ProcedureCache].
+    pub fn insert(&mut self, proc: Procedure) {
+        // If the `Entry` is `Occupied then do nothing as the Procedure is already in the cache. If
+        // the entry is `Vacant` then insert the Procedure.
+        if let Entry::Vacant(entry) = self.proc_id_map.entry(*proc.id()) {
+            let mast_hash = proc.code_root().hash();
+            entry.insert(mast_hash);
+            self.mast_map.entry(mast_hash).or_insert(proc);
+        }
```

Comment:
We could probably change this to something like:
```Rust
self.proc_id_map.entry(*proc.id()).or_insert_with(|| {
   ...
});
```
This way we don't need to import `Entry`.

Also, I wonder if we should handle insertions into these two maps separately. I'm thinking about the situation where we'll have procedures without a meaningful ID (e.g., for `call.0x1234567` case). These procedures would still need to be added to the MAST |-> Procedure map, but not to ProcedureId |-> Procedure map. What do you think?

Reply:
> We could probably change this to something like:
> 
> ```rust
> self.proc_id_map.entry(*proc.id()).or_insert_with(|| {
>    ...
> });
> ```

Good catch I'll try this.

> Also, I wonder if we should handle insertions into these two maps separately. I'm thinking about the situation where we'll have procedures without a meaningful ID (e.g., for `call.0x1234567` case). These procedures would still need to be added to the MAST |-> Procedure map, but not to ProcedureId |-> Procedure map. What do you think?

How would you define a `Procedure` without a `ProcedureId`?  Typically we use the `proc`/`export` keyword followed by the procedure name.  Do you envision some other mechanic for defining procedures and inserting them in the cache? One thing I'm not sure / happy about is the case when we have two procedures with the same mast root.  Lets take the following example:

```
export.proc_1
     push.1.2.3.4
end

export.proc_2
    push.1.2.3.4
end
```  

This would result in the following `ProcedureCache`:

```
ProcedureCache {
      proc_id_map: [proc_1_id -> MAST, proc_2_id -> MAST],
      mast_map: [MAST -> Procedure { id: proc_1_id, label: proc_1, code_root ...}] 
}
```

The mast_map only contains a single entry as both procedures have the same MAST root.  This means that when the client requests the procedure associated with proc_2_id it will actually return the procedure associated with proc_1.  The data model here isn't suitable. Upon reflection maybe it is preferable to have a data structure as follows:

```rust
pub struct ProcedureCache {
    proc_id_map: BTreeMap<ProcedureId, Procedure>,
    mast_map: BTreeMap<RpoDigest, ProcedureId>,
}
```

Reply:
> How would you define a `Procedure` without a `ProcedureId`?

We can't really but I was thinking maybe we'd end up with some ID collisions if, for example, we compile 2 account modules under common namespace. Let's say both accounts export `receive_assets` and the namespace we use is `context/account`. If implementation of these procedures is the same (i.e., same MAST root) then it doesn't matter - but if it is different, we'd lose the second procedure.

I think for now we could just make sure that ID collisions trigger some kind of error (it would be on the user of the assembler to make sure such errors don't happen) and then the new structure procedure cache (with `proc_id_map` and `mast_map`) should work.

In the future, I wonder if we should introduce 2 procedure variants - e.g., `NamedProcedure` and `AnonymousProcedure`.  Then, when compiling account modules we could somehow indicate that all top-level procedures should be anonymous. There could be other benefits to having this separation - but this requires more thinking.

Reply:
> We can't really but I was thinking maybe we'd end up with some ID collisions if, for example, we compile 2 account modules under common namespace. Let's say both accounts export `receive_assets` and the namespace we use is `context/account`. If implementation of these procedures is the same (i.e., same MAST root) then it doesn't matter - but if it is different, we'd lose the second procedure.

Do we want to support compilation of multiple modules under the same namespace and if so in what cases?

> I think for now we could just make sure that ID collisions trigger some kind of error (it would be on the user of the assembler to make sure such errors don't happen) and then the new structure procedure cache (with `proc_id_map` and `mast_map`) should work.

I have updated the PR to raise an error on collisions as advised.

> In the future, I wonder if we should introduce 2 procedure variants - e.g., `NamedProcedure` and `AnonymousProcedure`. Then, when compiling account modules we could somehow indicate that all top-level procedures should be anonymous. There could be other benefits to having this separation - but this requires more thinking.

Yes I think this makes sense.  I also wonder if we want to split the `Procedure` implementation in such a way that decouples the procedure body from the procedure metadata.  We can have many procedures that have the same procedure body but different metadata (e.g. `ProcedureId` / `ProcedureName` etc).  We would then have a many to one relationship from procedure metadata -> procedure body related via the mast root hash.  The procedure metadata could then be an enum of `MastProcedure` and `NamedProcedure`.

Reply:
> Do we want to support compilation of multiple modules under the same namespace and if so in what cases?

I'm thinking here more about how we'd compile multiple account modules for the same client. For example, say I have a wallet with 2 accounts. These accounts export an overlapping set of procedures (i..e., some procedures have the same names in both accounts, but the MAST roots may differ).

When we load these accounts into the assembler (via `compile_module`). We could use the same namespace (e.g., `context/account` - but then procedures with the same name would resolve to the same ID and it would be great to get an error at this point. A better thing is probably to compile them under different namespaces (e.g., `context/account0x12345` and `context/account0x54321`) - this way, even if there are procedure name collisions we'd get different procedure IDs.

In the long term, I think an even better solution would be to compile account modules into anonymous procedures

> I also wonder if we want to split the `Procedure` implementation in such a way that decouples the procedure body from the procedure metadata. We can have many procedures that have the same procedure body but different metadata (e.g. `ProcedureId` / `ProcedureName` etc). We would then have a many to one relationship from procedure metadata -> procedure body related via the mast root hash.

Excellent suggestion! Yes, I think we should definitely do something like this in a future PR.

Reply:
> When we load these accounts into the assembler (via `compile_module`). We could use the same namespace (e.g., `context/account` - but then procedures with the same name would resolve to the same ID and it would be great to get an error at this point. A better thing is probably to compile them under different namespaces (e.g., `context/account0x12345` and `context/account0x54321`) - this way, even if there are procedure name collisions we'd get different procedure IDs.

Makes sense. I think compiling them under different namespaces would be preferable. 

> In the long term, I think an even better solution would be to compile account modules into anonymous procedures.

Yes I think the introduction of anonymous procedures / mast procedures makes sense.  The way I am thinking about this is the `MastProcedure` object would still retain the scope information through the `ProcedureId` and be uniquely identified by it's mast root.  Something like

```rust
MastProcedure {
     root: RpoDigest,
     id: ProcedureId
}
``` 

where id = ProcedureId::new("library::module::0x123456...").

The concept of an anonymous invocation would be based on the way the calling masm is defined - it is absent of module scope information as seen below:

```
call.0x123456...
```

This invocation instructs that we do not need any guarantees about the source of the invocation target.

However, in some cases as we have discussed we may want additional guarantees and in those instances we can use the path to provide those guarantees:

```
use.library::module

...

call.module::0x123456...
```

Reply:
> ```rust
> MastProcedure {
>      root: RpoDigest,
>      id: ProcedureId
> }
> ```

Possibly - though, I'm wondering if there is a way we can get away with just MAST root.

> However, in some cases as we have discussed we may want additional guarantees and in those instances we can use the path to provide those guarantees:
> 
> ```
> use.library::module
> 
> ...
> 
> call.module::0x123456...
> ```

My current preference would be to embed additional info into the MAST call itself. For this, we could extend MAST call syntax with an extra (arbitrary?) metadata like so:
```
call.0x1234567@metadata
```
But this still needs some discussion.